### DISCUSSION TITLE
How to get some nodes like the Roslyn syntax visualizer in Visual Studio?

### DISCUSSION THREAD

**Discussion by wtujvk** [dotnet/roslyn#55934] [Q&A]
I want to get namespace of Qualified Name, so use roslyn.
`
 var filePath1 = "D:\\temp0\\logs\\2021-05-24-Error.log";
            System.String extension = System.IO.Path.GetExtension(filePath1);
            Action abc = () => System.IO.Path.GetFileName(System.IO.Path.GetExtension("xxy.log"));
`
![1](https://user-images.githubusercontent.com/17868266/128016417-aa2df793-f626-48e9-8444-c2f2ee3dbc03.png)
![2](https://user-images.githubusercontent.com/17868266/128016431-0c78cd13-679b-4862-b83c-09f611fef0f9.png)
![3](https://user-images.githubusercontent.com/17868266/128016444-f3dc574f-d107-4f26-abab-5d6a90ae17c7.png)

I use code like this:
`var code = @"var txtContent = DateTime.Now.ToString();
txtContent = System.IO.File.ReadAllText('2021-05-24-Error.log', Encoding.UTF8)";`
`SyntaxTree tree = CSharpSyntaxTree.ParseText(code);
CompilationUnitSyntax root = tree.GetCompilationUnitRoot();
var members = root.Members;   // one member one sentence.
 foreach (var item in members){
 // ...
}
`
What next , `item.DescendantNodes()`, `item.DescendantNodesAndTokens()`, `item.DescendantTrivia()`, how to get Classification ?
I fand the const value in Microsoft.CodeAnalysis.Classification.ClassificationTypeNames

Environment and configuration:
`  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp3.1</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.CodeAnalysis.CSharp.Workspaces" Version="3.10.0" />
    <PackageReference Include="Newtonsoft.Json" Version="13.0.1" />
  </ItemGroup>`

**Comment by Youssef1313** [comment]
Not 100% sure, but it's probably `IClassificationTypeRegistryService`. https://docs.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.text.classification.iclassificationtyperegistryservice?view=visualstudiosdk-2019

**Comment by Youssef1313** [comment]
or probably it's [`Classifier.GetClassifiedSpansAsync`](https://docs.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.classification.classifier.getclassifiedspansasync?view=roslyn-dotnet-3.10.0)

**Comment by wtujvk** [comment]
I get the ClassificationType is all 'identifier' otherwise  'operator'.
code string
`var code = DateTime.Now.ToString();
txtContent = System.IO.File.ReadAllText('2021-05-24-Error.log', Encoding.UTF8)`

Get Document :
`        private static Document GetDocument(string code)
        {
            var workspace = new AdhocWorkspace();

            string projectName = "HelloWorldProject";
            ProjectId projectId = ProjectId.CreateNewId();
            VersionStamp versionStamp = VersionStamp.Create();
            ProjectInfo helloWorldProject = ProjectInfo.Create(projectId, versionStamp, projectName, projectName, LanguageNames.CSharp);
            SourceText sourceText = SourceText.From(code);

            Project newProject = workspace.AddProject(helloWorldProject);
            Document newDocument = workspace.AddDocument(newProject.Id, "NewFile.cs", sourceText);
            return newDocument;
        }`

Get member:
`var document = GetDocument(code);
            var tree = document.GetSyntaxTreeAsync().ConfigureAwait(false).GetAwaiter().GetResult();
            var root = tree.GetCompilationUnitRoot();
            var members = root.Members;
            foreach (var item in members){
     
}`

Print value:
` private static IEnumerable<string> FindUsedUsingsFromSyntaxNode2(IEnumerable<SyntaxNode> nodes, Document document)
        {
            var classifiedSpans = nodes.Select(x => new {
                ClassifiedSpans = Classifier.GetClassifiedSpansAsync(document, x.FullSpan).ConfigureAwait(false).GetAwaiter().GetResult(),
                FullSpan = x.FullSpan,
                FullString = x.ToFullString()
            }).ToList();
            foreach (var classifiedSpansItem in classifiedSpans)
            {
                Console.WriteLine();
                Console.WriteLine($"{classifiedSpansItem.FullString}:");
                foreach (var classifiedSpanDetailItem in classifiedSpansItem.ClassifiedSpans)
                {
                    Console.WriteLine($"{classifiedSpanDetailItem.ClassificationType}");
                }
            }

            ////var searchClassifiedSpans = classifiedSpans.Where(c => c.ClassifiedSpans.Any(d => d.ClassificationType == ClassificationTypeNames.ClassName || d.ClassificationType == ClassificationTypeNames.StaticSymbol));

            return new List<string>();
        }`


The expectation is 'Namespace'、‘Static Symbol’,  but it  get 'identifier' or 'operator'.

![1](https://user-images.githubusercontent.com/17868266/128108848-ff4c3dc1-47ca-49da-8fd2-b1caeaf971a8.jpg)
![2](https://user-images.githubusercontent.com/17868266/128108852-628a0b83-9367-42e9-8494-d5a9e180dddd.jpg)


So, what error ?
